In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import math
import itertools
from matplotlib.ticker import ScalarFormatter
pd.set_option('display.max_columns', None)

In [2]:
df_raw = pd.read_csv("dropout_rate.csv", sep=";", encoding="utf-8-sig")

In [3]:
df_raw

Marital status  Application mode  Application order  Course  \
0                  1                17                  5     171   
1                  1                15                  1    9254   
2                  1                 1                  5    9070   
3                  1                17                  2    9773   
4                  2                39                  1    8014   
...              ...               ...                ...     ...   
4419               1                 1                  6    9773   
4420               1                 1                  2    9773   
4421               1                 1                  1    9500   
4422               1                 1                  1    9147   
4423               1                10                  1    9773   

      Daytime/evening attendance\t  Previous qualification  \
0                                1                       1   
1                                1                       1   
2                                1                       1   
3                                1                       1   
4                                0                       1   
...                            ...                     ...   
4419                             1                       1   
4420                             1                       1   
4421                             1                       1   
4422                             1                       1   
4423                             1                       1   

      Previous qualification (grade)  Nacionality  Mother's qualification  \
0                              122.0            1                      19   
1                              160.0            1                       1   
2                              122.0            1                      37   
3                              122.0            1                      38   
4                              100.0            1                      37   
...                              ...          ...                     ...   
4419                           125.0            1                       1   
4420                           120.0          105                       1   
4421                           154.0            1                      37   
4422                           180.0            1                      37   
4423                           152.0           22                      38   

      Father's qualification  Mother's occupation  Father's occupation  \
0                         12                    5                    9   
1                          3                    3                    3   
2                         37                    9                    9   
3                         37                    5                    3   
4                         38                    9                    9   
...                      ...                  ...                  ...   
4419                       1                    5                    4   
4420                       1                    9                    9   
4421                      37                    9                    9   
4422                      37                    7                    4   
4423                      37                    5                    9   

      Admission grade  Displaced  Educational special needs  Debtor  \
0               127.3          1                          0       0   
1               142.5          1                          0       0   
2               124.8          1                          0       0   
3               119.6          1                          0       0   
4               141.5          0                          0       0   
...               ...        ...                        ...     ...   
4419            122.2          0                          0       0   
4420            119.0          1                  

## Mô tả

Bộ dữ liệu được tạo ra nhằm hỗ trợ giảm tỷ lệ bỏ học và thất bại học tập ở bậc đại học, thông qua việc ứng dụng các kỹ thuật học máy để xác định sinh viên có nguy cơ cao ngay từ giai đoạn đầu của quá trình học. Bộ dữ liệu tạo bởi nhóm Mónica V. Martins, Daniel Tolledo, Jorge Machado, Luís M. T. Baptista, and Valentim Realinho. 2021

In [4]:
df_raw.shape

(4424, 37)

In [5]:
df_raw.isnull().sum()

Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            0
Daytime/evening attendance\t                      0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            0
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   0
Displaced                                         0
Educational special needs                         0
Debtor                                            0
Tuition fees up to date                           0
Gender                                            0
Scholarship holder                                0
Age at enrol

Số dòng: 4,424

Số cột: 36

Không có giá trị thiếu (ko null).

# Phân loại

## Qualitative
| Cột | Ý nghĩa |
|---|---|
| Marital Status | Tình trạng hôn nhân: 1 – độc thân, 2 – đã kết hôn, 3 – góa, 4 – ly hôn, 5 – ly thân hợp pháp |
| Application mode | Hình thức nộp hồ sơ (ví dụ: kỳ 1, diện đặc biệt, ưu tiên vùng...) |
| Application order | Mức độ ưu tiên của ngành khi đăng ký (0 là ưu tiên cao nhất) |
| Course | Mã ngành học (ví dụ: 33 - Công nghệ nhiên liệu sinh học, 171 - Đồ họa hoạt hình...) |
| Daytime/evening attendance | Thời gian học: 1 – ban ngày, 0 – buổi tối |
| Previous qualification | Trình độ học vấn trước đó |
| Nacionality | Quốc tịch (được mã hóa) |
| Mother's qualification | Trình độ học vấn của mẹ |
| Father's qualification | Trình độ học vấn của cha |
| Mother's occupation | Nghề nghiệp của mẹ |
| Father's occupation | Nghề nghiệp của cha |
| Displaced | Có phải sống xa nhà không: 1 - Có, 0 - Không |
| Educational special needs | Có nhu cầu giáo dục đặc biệt không: 1 - Có, 0 - Không |
| Debtor | Có nợ học phí không: 1 - Có, 0 - Không |
| Tuition fees up to date | Đã đóng đủ học phí chưa: 1 - Rồi, 0 - Chưa |
| Gender | Giới tính: 1 - Nam, 0 - Nữ |
| Scholarship holder | Có học bổng không: 1 - Có, 0 - Không |
| International | Sinh viên quốc tế: 1 - Có, 0 - Không |
| Target | Kết quả học tập: "Graduate" (tốt nghiệp), "Dropout" (bỏ học), ... |

## Quantitative 

### Discrete

| Cột | Ý nghĩa |
|---|---|
| Age at enrollment | Tuổi khi nhập học |
| Curricular units 1st sem (credited) | Số học phần kỳ 1 được miễn/chuyển điểm |
| Curricular units 1st sem (enrolled) | Số học phần kỳ 1 đã đăng ký |
| Curricular units 1st sem (evaluations) | Số học phần kỳ 1 đã tham gia đánh giá |
| Curricular units 1st sem (approved) | Số học phần kỳ 1 đã qua môn |
| Curricular units 1st sem (without evaluations) | Số học phần kỳ 1 không có điểm đánh giá |
| Curricular units 2nd sem (credited) | Số học phần kỳ 2 được miễn/chuyển điểm |
| Curricular units 2nd sem (enrolled) | Số học phần kỳ 2 đã đăng ký |
| Curricular units 2nd sem (evaluations) | Số học phần kỳ 2 đã tham gia đánh giá |
| Curricular units 2nd sem (approved) | Số học phần kỳ 2 đã qua môn |
| Curricular units 2nd sem (without evaluations) | Số học phần kỳ 2 không có điểm đánh giá |

### Continuous

| Cột | Ý nghĩa |
|---|---|
| Previous qualification (grade) | Điểm số bằng cấp trước đó |
| Admission grade | Điểm đầu vào |
| Curricular units 1st sem (grade) | Điểm trung bình kỳ 1 |
| Curricular units 2nd sem (grade) | Điểm trung bình kỳ 2 |
| Unemployment rate | Tỉ lệ thất nghiệp tại thời điểm đó |
| Inflation rate | Tỉ lệ lạm phát tại thời điểm đó |
| GDP | Tổng sản phẩm quốc nội tại thời điểm đó |


In [8]:
df = df_raw.copy()

summary_stats = {
    "Thông số thống kê": [
        "Mean",
        "Min",
        "25th Percentile",
        "Median",
        "75th Percentile",
        "Max",
        "Mode",
        "Variance",
        "Standard Deviation",
        "Coefficient of Variation"
    ]
}

numeric_discrete = [
    "Age at enrollment",
    "Curricular units 1st sem (credited)",
    "Curricular units 1st sem (enrolled)",
    "Curricular units 1st sem (evaluations)",
    "Curricular units 1st sem (approved)",
    "Curricular units 1st sem (without evaluations)",
    "Curricular units 2nd sem (credited)",
    "Curricular units 2nd sem (enrolled)",
    "Curricular units 2nd sem (evaluations)",
    "Curricular units 2nd sem (approved)",
    "Curricular units 2nd sem (without evaluations)"
]


for col in numeric_discrete:
    data = df[col]
    mean_val = data.mean()
    std_val = data.std()
    mode_val = data.mode().iloc[0] if not data.mode().empty else np.nan
    variance_val = data.var()
    cv_val = std_val / mean_val if mean_val != 0 else np.nan

    summary_stats[col] = [
        mean_val,
        data.min(),
        data.quantile(0.25),
        data.median(),
        data.quantile(0.75),
        data.max(),
        mode_val,
        variance_val,
        std_val,
        cv_val
    ]

formatted_summary_df = pd.DataFrame(summary_stats).copy()

for col in formatted_summary_df.columns[1:]:  # Skip "Thông số thống kê"
    formatted_summary_df[col] = formatted_summary_df[col].apply(lambda x: f'{x:,.4f}' if isinstance(x, (int, float)) else x)

formatted_summary_df

Thông số thống kê Age at enrollment  \
0                      Mean           23.2651   
1                       Min           17.0000   
2           25th Percentile           19.0000   
3                    Median           20.0000   
4           75th Percentile           25.0000   
5                       Max           70.0000   
6                      Mode           18.0000   
7                  Variance           57.5749   
8        Standard Deviation            7.5878   
9  Coefficient of Variation            0.3261   

  Curricular units 1st sem (credited) Curricular units 1st sem (enrolled)  \
0                              0.7100                              6.2706   
1                              0.0000                              0.0000   
2                              0.0000                              5.0000   
3                              0.0000                              6.0000   
4                              0.0000                              7.0000   
5                             20.0000                             26.0000   
6                              0.0000                              6.0000   
7                              5.5720                              6.1513   
8                              2.3605                              2.4802   
9                              3.3247                              0.3955   

  Curricular units 1st sem (evaluations) Curricular units 1st sem (approved)  \
0                                 8.2991                              4.7066   
1                                 0.0000                              0.0000   
2                                 6.0000                              3.0000   
3                                 8.0000                              5.0000   
4                                10.0000                              6.0000   
5                                45.0000                             26.0000   
6                                 8.0000                              6.0000   
7                                17.4649                              9.5743   
8                                 4.1791                              3.0942   
9                                 0.5036                              0.6574   

  Curricular units 1st sem (without evaluations)  \
0                                         0.1377   
1                                         0.0000   
2                                         0.0000   
3                                         0.0000   
4                                         0.0000   
5                                        12.0000   
6                                         0.0000   
7                                         0.4773   
8                                         0.6909   
9                                         5.0188   

  Curricular units 2nd sem (credited) Curricular units 2nd sem (enrolled)  \
0                              0.5418                              6.2321   
1                              0.0000                              0.0000   
2                              0.0000                              5.0000   
3                              0.0000                              6.0000   
4                              0.0000                              7.0000   
5                             19.0000                             23.0000   
6                              0.0000                              6.0000   
7                              3.6808                              4.8222   
8                              1.9185                              2.1960   
9                              3.5409                              0.3524   

  Curricular units 2nd sem (evaluations) Curricular units 2nd sem (approved)  \
0                                 8.0633                              4.4358   
1                                 0.0000                              0.0000   
2                                 6.0000                              2.0000   
